meerk40t/extra/param_functions.py

```diff
@@ -0,0 +1,757 @@
+                0,
+                bb[0] + sdx,
+                bb[1],
+                0,
+                bb[2],
+                bb[1] + sdy,
+                1,
+                cols,
+                1,
+                rows,
+            )
+            node.altered()
+
+        # --- Fractal Tree
+        def create_fractal_tree(first_pt, second_pt, iterations, ratio):
+            def tree_fractal(geom, startpt, endpt, depth, ratio):
+                #
+                # create a line from startpt to endpt and add it to the geometry
+                if depth < 0:
+                    return
+                angle = startpt.angle_to(endpt)
+                dist = startpt.distance_to(endpt)
+                delta = math.tau / 8
+                # print(
+                #     f"{depth}: ({startpt.x:.0f}, {startpt.y:.0f}) - ({endpt.x:.0f}, {endpt.y:.0f}) : {dist:.0f}"
+                # )
+                geom.line(complex(startpt.x, startpt.y), complex(endpt.x, endpt.y))
+                newstart = Point(endpt)
+                newend = Point.polar(endpt, angle - delta, dist * ratio)
+                tree_fractal(geom, newstart, newend, depth - 1, ratio)
+                newend = Point.polar(endpt, angle + delta, dist * ratio)
+                tree_fractal(geom, newstart, newend, depth - 1, ratio)
+
+            geometry = Geomstr()
+            # print("create fractal", first_pt, second_pt)
+            tree_fractal(geometry, first_pt, second_pt, iterations, ratio)
+            return geometry
+
+        @self.console_argument("sx", type=Length)
+        @self.console_argument("sy", type=Length)
+        @self.console_argument("branch", type=Length)
+        @self.console_argument("iterations", type=int)
+        @context.console_command(
+            "fractal_tree", help=_("fractal_tree sx, sy, branch, iterations")
+        )
+        def fractal_tree(
+            command,
+            channel,
+            _,
+            sx=None,
+            sy=None,
+            branch=None,
+            iterations=None,
+            data=None,
+            post=None,
+            **kwargs,
+        ):
+            ratio = 0.75
+            try:
+                if sx is None:
+                    sx = Length("5cm")
+                ssx = float(sx)
+                if sy is None:
+                    sy = Length("5cm")
+                ssy = float(sy)
+                if branch is None:
+                    branch = Length("4cm")
+                blen = float(branch)
+                if iterations is None:
+                    iterations = 10
+                iterations = int(iterations)
+            except ValueError:
+                channel("Invalid data provided")
+                return
+            start_pt = Point(ssx, ssy)
+            end_pt = Point.polar(start_pt, 0, blen)
+            geom = create_fractal_tree(start_pt, end_pt, iterations, ratio)
+            node = self.elem_branch.add(type="elem path", geometry=geom)
```

Comment:
Ah, this is why the parameters aren't inherent to the code, since it's creating these things as paths and then massaging the functions to be sort of equal. --- This feels like these should be `elem fractal`.